⬆️ UPGRADE: sphinx v3, myst-nb v0.10 (#911)
executablebooks/jupyter-book: docs/content/figures.md

````md
```{image} ../images/fun-fish.png
:alt: fishy
:class: bg-primary
:width: 200px
:align: center
```
````

```{image} ../images/fun-fish.png
:alt: fishy
:class: bg-primary mb-1
:width: 200px
```

This allows you to control aspects of the image with [directive arguments](directive-arguments).

In one way, this is an improvement on the Markdown syntax, however, the drawback is that this syntax will not show the image in common Markdown viewers (for example when the files are viewed on GitHub).
The final option then is directly using HTML, which can also parsed by MyST Markdown.
U
sing raw HTML is usually a bad choice (see [this explanation](raw-html-in-markdown)),
but enabling extended syntax in your `_config` enables MySt-Parser to properly handle isolated `img` tags:

```yaml
parse:
  myst_extended_syntax: true
```

Now you can add:

```html
<img src="../images/fun-fish.png" alt="fishy" class="bg-primary" width="200px">
```

and we correctly render:

<img src="../images/fun-fish.png" alt="fishy" class="bg-primary mb-1" width="200px">

> This will also be output in PDF LaTeX builds!

Reviewer: @chrisjsewell no way! -- nicely done. That's cool. Is this in the parser to unpack `html`?

Author: yep also, as of today, see https://myst-parser.readthedocs.io/en/latest/using/syntax-optional.html#markdown-figures, which allows you to also add a reference target and a caption 😃

Author: This is the HTML parser code: https://github.com/executablebooks/MyST-Parser/blob/master/myst_parser/parse_html.py
actually pretty simple

Author: the `:::` blocks are basically a new "breed" of Markdown specific directives.
As I plan to add more (see https://github.com/executablebooks/MyST-Parser/issues/213), I will try to formalise their usage, so that they are more modular and extensible

# Images and Figures

(content-blocks-images)=
## Images

MyST Markdown provides a few different syntaxes for including images in your documentation, as explained below.

The first is the standard Markdown syntax:

```md
![fishy](../images/fun-fish.png)
```

![fishy](../images/fun-fish.png)

This will correctly copy the image to the build folder and will render it in all output formats (HTML, TeX, etc).
However, it is limited in the configuration that can be applied, for example setting a width.

> As discussed in [this section](content:myst/directives), MyST allow for directives to be used such as `image` and `figure` (see [the sphinx documentation](sphinx:rst-primer) for available options):

Reviewer 2: ```suggestion
As discussed in [this section](content:myst/directives), MyST allows for directives to be used such as `image` and `figure` (see [the sphinx documentation](sphinx:rst-primer) for available options):
```

````md
```{image} ../images/fun-fish.png
:alt: fishy
:class: bg-primary
:width: 200px
:align: center
```
````

```{image} ../images/fun-fish.png
:alt: fishy
:class: bg-primary mb-1
:width: 200px
```

This allows you to control aspects of the image with [directive arguments](directive-arguments).

In one way, this is an improvement on the Markdown syntax, however, the drawback is that this syntax will not show the image in common Markdown viewers (for example when the files are viewed on GitHub).

> The final option then is directly using HTML, which can also parsed by MyST Markdown.

Reviewer 2: ```suggestion
The final option then is directly using HTML, which can also be parsed by MyST Markdown.
```

> U

Reviewer 2: accidental line break?

> U
> sing raw HTML is usually a bad choice (see [this explanation](raw-html-in-markdown)),

Author: ```suggestion
Using raw HTML is usually a bad choice (see [this explanation](raw-html-in-markdown)),
```

with equivalent behaviour to the `figure` directive above.

To enable them, add in your `_config.yml`:

```yaml
parse:
  myst_extended_syntax: true
```

The figure block must contain **only** two components; an image, in either Markdown or HTML syntax, and a single paragraph for the caption.

As with admonitions, the figure can have additional classes set on it, but the title is now taken as the reference target of the figure:

```md
:::{figure,myclass} markdown-fig
<img src="../images/fun-fish.png" alt="fishy" class="bg-primary mb-1" width="200px">

This is a caption in **Markdown**!
:::
```

:::{figure,myclass} markdown-fig
<img src="../images/fun-fish.png" alt="fishy" class="bg-primary mb-1" width="200px">

This is a caption in **Markdown**!
:::

As we see here, the target we set can be referenced:

```md
[Go to the fish!](markdown-fig)
```

[Go to the fish!](markdown-fig)

## Referencing figures

You can then refer to your figures using the `{ref}` role or Markdown style references like:

```md
- {ref}`directive-fig`
- [](markdown-fig)
```

which will replace the reference with the figure caption like so:

- {ref}`directive-fig`
- [](markdown-fig)

> Another convenient way to create cross-references is with the `{numref}` role,

Reviewer 2: I think we should have a header that has "numbered references" or something, so people find it more easily

Author: added in coming commit
